pkg/k8s/resource/resource.go

```diff
@@ -469,6 +519,39 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 	return out
 }
 
+func (r *resource[T]) release() {
+	if !r.opts.releasable {
+		return
+	}
+
+	// in case of a releasable resource, stop the underlying informer when the last
+	// reference to it is released. The resource is restarted to be
+	// ready again in case of a subsequent call to either Events() or Store().
+
+	r.refsMu.Lock()
+	defer r.refsMu.Unlock()
+
+	r.refs--
+	if r.refs > 0 {
+		return
+	}
+
+	r.resetCancel()
```

Comment:
Do we need two separate contexts or could we just use r.ctx and r.cancel?

Reply:
I think we need to separate the two signals, as the existing `r.ctx` is meant to stop the resource once and for all, while the new one aims to restart it. If we want to stop the resource definitively (i.e: hive is stopping), we shouldn't restart the goroutine in `startWhenNeeded`, otherwise we'll leak it.